### ISSUE TITLE
Thrown exception in `nothrow` function not detected

### ISSUE THREAD

**Issue by isuckatcs** [llvm/llvm-project#60086]
If an exception is thrown in a function marked with `nothrow` clang usually reports a warning.

In the following example however an exception is thrown but clang fails to detect it and the 
warning is not reported at all.

```c++
void foo() noexcept {
  try {
    double *a[2][3];
    throw a;
  } catch (const double *const(*)[3]) {
  }
}
```

What makes this more interesting is that on [cppreference](https://en.cppreference.com/w/cpp/language/try_catch) the following can be seen:
```
When an exception is thrown by any statement in compound-statement, the exception object of type E is 
matched against the types of the formal parameters T of each catch-clause in handler-seq, in the order in 
which the catch clauses are listed. The exception is a match if any of the following is true:

...
- T is (possibly cv-qualified) U or const U& (since C++14), and U is a pointer or pointer to member type, and
  E is also a pointer or pointer to member type that is implicitly convertible to U by one or more of
  - a standard pointer conversion other than one to a private, protected, or ambiguous base class
  - **a qualification conversion**
  - a function pointer conversion (since C++17)
...
```
While the [qualification conversion section](https://en.cppreference.com/w/cpp/language/implicit_conversion#Qualification_conversions) lists a number of rules that describe when can 
the said conversion be performed, it gives the following example:
```c++
double *a[2][3];
double const * const (*ap)[3] = a; // OK
```
So technically the thrown type in `foo()` can be converted to the type in the handler, which
is the reason for clang not reporting a warning but the exception is thrown regardless. 

See the example on [godbolt](https://godbolt.org/z/sEz8eYeP1).
